packages/neuron-wallet/src/services/addresses.ts

```diff
@@ -149,128 +173,194 @@ export default class AddressService {
+        hashType: SystemScriptInfo.SECP_HASH_TYPE,
+        args: address.blake160
+      });
+      const lockHash = script.computeHash()
+      const txCount = txCountsByLockHashes.get(lockHash)
+
+      if (!txCount) {
+        continue
+      }
+      address.txCount = txCount
     }
-    return addressEntity
-  }
 
-  public static allAddresses = (): AddressInterface[] => {
-    return AddressDao.allAddresses( AddressService.getAddressVersion())
-  }
+    const allUnusedAddresses = addresses.filter(address => !address.txCount)
 
-  public static allAddressesByWalletId = (
-    walletId: string,
-    addressVersion: AddressVersion = AddressService.getAddressVersion()
-  ): AddressInterface[] => {
-    return AddressDao.allAddressesByWalletId(walletId, addressVersion)
+    return allUnusedAddresses
   }
 
-  public static allBlake160sByWalletId(walletId: string): string[] {
-    return AddressService.allAddressesByWalletId(walletId).map(addr => addr.blake160)
+  public static async getNextUnusedAddressByWalletId (walletId: string): Promise<AddressInterface | undefined> {
+    const [unusedReceivingAddresses, unusedChangeAddresses] = await this.getGroupedUnusedAddressesByWalletId(walletId)
+    if (unusedReceivingAddresses.length) {
+      return unusedReceivingAddresses[0]
+    }
+    if (unusedChangeAddresses.length) {
+      return unusedChangeAddresses[0]
+    }
+
+    return undefined
   }
 
-  public static allLockHashes(): string[] {
-    const addresses = AddressService.allAddresses().map(address => address.address)
-    return AddressParser.batchToLockHash(addresses)
+  public static async getNextUnusedChangeAddressByWalletId (walletId: string): Promise<AddressInterface | undefined> {
+    const [ , unusedChangeAddresses ] = await this.getGroupedUnusedAddressesByWalletId(walletId)
+    if (unusedChangeAddresses.length) {
+      return unusedChangeAddresses[0]
+    }
+    return undefined
   }
 
-  public static allLockHashesByWalletId(walletId: string): string[] {
-    const addresses = AddressService.allAddressesByWalletId(walletId).map(addr => addr.address)
-    return AddressParser.batchToLockHash(addresses)
+  public static async getUnusedReceivingAddressesByWalletId(walletId: string): Promise<AddressInterface[]> {
+    const [ unusedReceivingAddresses ] = await this.getGroupedUnusedAddressesByWalletId(walletId)
+    return unusedReceivingAddresses
   }
 
-  public static usedAddresses = (walletId: string): AddressInterface[] => {
-    return AddressDao.usedAddressesByWalletId(walletId,  AddressService.getAddressVersion())
+
+  public static getAddressesByAllWallets = async (): Promise<AddressInterface[]> => {
+    const publicKeyInfos = await getConnection()
+      .getRepository(HdPublicKeyInfo)
+      .createQueryBuilder()
+      .getMany()
+
+    return publicKeyInfos
+      .map(publicKeyInfo => publicKeyInfo.toModel())
+      .map(model => (AddressMeta.fromHdPublicKeyInfoModel(model)))
```

Comment:
Is there an extra loop?  We only need to call the `map` function once to get the result.